Introduce CMP library only

GHaberis opened the pull request:
## What does this change?

I've pulled out `lib/cmp` and `lib/cmp.spec.js` into this PR from the original https://github.com/guardian/frontend/pull/21696. The module `lib/cmp.js` is initialised in `standard` in preparation for it to be used it, although it isn't used yet - subsequent PRs for this integration work to follow.

PRBuilds commented:
PRbuilds results:


**Screenshots**
[wide.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-21705/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/wide.png) • [desktop.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-21705/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/desktop.png) • [tablet.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-21705/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/tablet.png) • [mobile.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-21705/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/mobile.png)







&#x1F49A; **A11y validation**
[a11y-report.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-21705/a11y-report.txt)





&#x1F49A; **Microdata Validation**
[microdata.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-21705/microdata.txt)



**Apache Benchmark Load Testing**
[loadtesting.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-21705/loadtesting.txt)



**LightHouse Reporting**



--automated message

adamnfish commented on static/src/javascripts/lib/cmp.spec.js:
```diff
@@ -0,0 +1,215 @@
+
+    describe('consentState', () => {
+        it('returns null state if cmp lib has not been initialised', () => {
+            expect(consentState('functional')).toBeNull();
```
This makes me wonder if the `consentState` functionality should even be exposed. Right now it's possible for a future developer that doesn't understand this code to just read the consent state from their library directly, without using the callbacks. In this case they'd likely just get `null` and might not know they need to use the listener approach to be told when the CMP library has loaded up.

What's the use case for exposing `consentState`?

ripecosta replied:
`consentState` is needed in such cases where asynchronous module isn't possible or parameter change isn't possible mid execution without major rework of the module's code. Such an example is the youtube player, whether the personalised ads flag is set when the player is started. Changes to the player can be made such that (for example) it reinitialises completely with the new state if a video isn't playing but that would mean a rework of that whole module, which is something we can't afford at the moment.  For this and other such cases we're sticking with checking the consent state at module init (same as it was before) and that's why we expose consentStat. Don't forget cmp lib's init is synchronous so everything after that will have the correct consent state already set in the lib.

Having said that this test's description is misleading. Perhaps it should be changed to something like "returns null if consent hasn't been set by the user" and the CMP lib README should clearly state that it should be run as early as possible to avoid module's calling it before it's ready.

GHaberis replied:
> this test's description is misleading. 

I disagree, it's possible to call `consentState` before the cmp lib has been initialised, and whilst this is possible we should understand how it'll behave, and this test demonstrates that. If you update this test to test "returns null if consent hasn't been set by the user" as suggested then it'd fail as right now it returns `true` for `functional` and `performance`.

ripecosta replied:
It's only misleading because it's not clear that even though we account for it calling `consentState()` before running `init()` is not the right way to use the library. Sorry for not being clear before.

ripecosta replied:
**Update**: Had a chat with @GHaberis and we decided to remove `init()` altogether to avoid confusion and make the library more robust. We'll make changes so the init logic is executed when a function from the library is first called.

adamnfish replied:
> We'll make changes so the init logic is executed when a function from the library is first called.
Much more sensible, nice one.

GHaberis replied:
> `consentState` is needed in such cases where asynchronous module isn't possible or parameter change isn't possible mid execution without major rework of the module's code. Such an example is the youtube player, whether the personalised ads flag is set when the player is started. 

We could  update the youtube initialisation step here https://github.com/guardian/frontend/blob/master/static/src/javascripts/bootstraps/enhanced/youtube.js#L6 so it's wrapped in a `onConsentNotification` call, and we pass the advertising consent state down like so:

```js
export const init = () => {
  onConsentNotification('advertisement', state => {
    checkElemsForVideos(state);
  });
};
```

Of course you'd only want to run `checkElemsForVideos` once. In which case you'd have to  add a flag of some sorts, as so:

```js
export const init = () => {
  let hasRun = false;
  onConsentNotification('advertisement', state => {
    if (!hasRun) {
      checkElemsForVideos(state);
      hasRun = true;
    }
  });
};
```

I also wonder whether  we want to build the idea of  only running once into the parameters of onConsentNotification,  for example add a third optional parameter  "runOnce"  (boolean, default = false). So you'd call it like:

```js
export const init = () => {
  onConsentNotification('advertisement', state => {
    checkElemsForVideos(state);
  }, true);
};
```

ripecosta replied:
I agree with your first suggestions to pass `checkElemsForVideos()` into a callback to `onConsentNotification()` but adding a runOnce boolean seems to go against what we want to do with consent, which is that each module should adapt every time consent is changed, not just the first time it changes.

ripecosta replied:
Apart from youtube only 2 other modules use `consentState()`. Those are `consent-tracker.js` and `user-ad-targeting.js`. The first we don't really need to worry about because it'll deleted when we start using the new consent audit log. The second can be converted to use a similar solution to the youtube one.

Which means that if no one can think of any other use cases we're safe to remove `consentState()`.
What do you think @GHaberis ?

GHaberis replied:
> I agree with your first suggestions to pass `checkElemsForVideos()` into a callback to `onConsentNotification()` but adding a runOnce boolean seems to go against what we want to do with consent, which is that each module should adapt every time consent is changed, not just the first time it changes.

The youtube initialisation is an example where we wouldn't want to adapt and run `checkElemsForVideos` on subsequent `onConsentNotification` events, these videos initialise once based on the first state given to them and that's it. In which case we have 2 options, the flag as suggested or some kind of `runOnce` parameter.

GHaberis replied:
> Apart from youtube only 2 other modules use `consentState()`. Those are `consent-tracker.js` and `user-ad-targeting.js`. The first we don't really need to worry about because it'll deleted when we start using the new consent audit log. The second can be converted to use a similar solution to the youtube one.
> 
> Which means that if no one can think of any other use cases we're safe to remove `consentState()`.
> What do you think @GHaberis ?

If we can remove for now then that sounds good to me 👍

adamnfish commented on static/src/javascripts/lib/cmp.spec.js:
```diff
@@ -0,0 +1,215 @@
+        it('returns functional consent state', () => {
+            init();
+
+            expect(consentState('functional')).toBe(true);
```
I still don't like these being hard-coded to true here, but I do see why that's the case. These tests will need to depend on the current state of the user's consent. Of course, right now we don't ask about functional so the current consent state is irrelevant, it just means these tests will be changing in the future. It's ok to leave this as is as long as we're careful to make sure this gets fixed later!

GHaberis replied:
This'll certainly be fixed in the code which will in turn cause this test to become out of date at which point we'll definitely need to fix the test.

adamnfish commented on static/src/javascripts/lib/cmp.spec.js:
```diff
@@ -0,0 +1,215 @@
+
+    describe('onConsentNotification', () => {
+        describe('if cmpIsReady is TRUE when onConsentNotification called', () => {
+            it('executes functional callback', () => {
```
This is really two tests, one that checks the execution and another that checks the value. `executes functional callback` is the `toHaveBeenCalledTimes(1)` check, and in this case it doesn't really matter what the value is.

Not a huge deal, but it's nice to have one assertion per test if possible.

GHaberis replied:
Happy to refactor suggested 👍

adamnfish commented on static/src/javascripts/lib/cmp.spec.js:
```diff
@@ -0,0 +1,215 @@
+                expect(myCallBack).toBeCalledWith(false);
+            });
+
+            it('executes advertisement callback each time consent nofication triggered', () => {
```
This would be a great candidate for property based testing, where given a random number, `n`, we check that `n` re-triggers leads to `n + 1` invocations. At the moment we're only checking it can be re-triggered once. @janua found a good library for this sort of thing in Typescript and may have tips here!

GHaberis replied:
Interesting idea, I'll have a go at implementing something that does this!

ripecosta replied:
That's a great idea!

adamnfish commented on static/src/javascripts/lib/cmp.spec.js:
```diff
@@ -0,0 +1,215 @@
+                    [true], // First call
+                    [true], // Second call
+                ]);
+            });
```
I don't see any tests to check that it can be re-triggered after a change to the user's consent. We'd want to make sure that the callbacks are invoked with the new value.

GHaberis replied:
The test `execute advertisement callback each time consent nofication triggered` does via the call `_.triggerConsentNotification();`, calling this simulates what should be executed after a change to the user's consent has been made. 

Currently this module only exposes `triggerConsentNotification` as  part of `_` for testing purposes. However I foresee `triggerConsentNotification` being on the public interface for the module once we implement the CMP UI, as the UI will need to call  `triggerConsentNotification` whenever the user submit's their consent via the UI and it has been saved in the cookie. 

We could just expose `triggerConsentNotification` now as part of the actual interface (not just for testing purposes) as we're going to need this exposed soon anyway. What do you think?

adamnfish replied:
I mean testing that the consent starts at `X`, then is changed to `Y`. The callbacks should be invoked with `X` and then `Y`. All the examples I spotted checked repeated invocations, but didn't test that the callback arguments can change between invocations.

Sorry for not being clear!

ripecosta replied:
@GHaberis triggerConsentNotification should only need to be exposed as a test function. Our saveCookie function will trigger it internally every time it's called.

@adamnfish As we don't yet have a way to change the consent state the current tests are the best we can do. But we'll add the tests you described when functions such as saveCookie and readCookie are implemented on the library.

GHaberis replied:
> As we don't yet have a way to change the consent state the current tests are the best we can do. But we'll add the tests you described when functions such as saveCookie and readCookie are implemented on the library.

This is correct, whilst there's no UI there's nothing to update the consent state right now.

> triggerConsentNotification should only need to be exposed as a test function. Our saveCookie function will trigger it internally every time it's called.

That depends on whether `saveCookie` and `triggerConsentNotification` are in the same module, I'd imagined they won't be as the scope of such a module would be quite broad, and prefer them to be in separate modules  with distinct purposes, in which case `triggerConsentNotification` would need to be exposed.

ripecosta replied:
I agree. But I didn't mean the function that saves the actual cookie. Just the externally facing function (whatever its name is) that causes a change in consent and subsequently triggers the cookie save. That one will most likely be in the same module as triggerConsentNotification and that's where the cookie state will be kept.

adamnfish replied:
> we don't yet have a way to change the consent state

There is an existing mechanism for the advertising purpose, but no worries if you want to introduce all of this together later.

> But we'll add the tests you described when functions such as saveCookie and readCookie are implemented on the library

Sounds good.

GHaberis commented:
@ripecosta @adamnfish I've made an update in which I removed `init` from the `lib/cmp` interface. I'd been thinking about `cmp.init` and how we're having to guard against it being called multiple times OR not called at all. Rather than guard against these situations I've removed the chance of them happening at all. We now only call `init` on the first `consentState` or `onConsentNotification` call, whichever comes first.

This simplifies the  interface, and reduces the need to have to call `cmp.init`.

ripecosta commented on static/src/javascripts/lib/cmp.js:
```diff
@@ -0,0 +1,89 @@
+): void => {
+    const purpose = purposes[purposeName];
+
+    if (!cmpIsReady) {
```
Should we move this logic into `init()` and simply call it in every function?
Do we need that same check on the test function `triggerConsentNotification()`? I think we probably do.

jeteve commented on static/src/javascripts/lib/cmp.js:
```diff
@@ -0,0 +1,85 @@
+const triggerConsentNotification = (): void => {
+    checkCmpReady();
+
+    Object.keys(purposes).forEach(key => {
```
Is that OK if purposes are triggered in random orders?

GHaberis replied:
We made a decision to keep this library naive to the execution order, it's the modules that pass the callbacks that should account for that.

jeteve replied:
Yes but here, each purpose will be triggered in an undefined order. I suppose that's ok :)

ripecosta replied:
That's right. There's no dependencies between modules in different purposes so the order we trigger these is irrelevant.
